Add bindToNetwork and unbindNetwork methods

edpaget opened the pull request:
Adds methods to force Android to use the connected Wifi network. Normally android will not use a wifi network that does not have internet connectivity if it has another network available (like cellular) that does.

tadasr commented:
@edpaget, this is a great propose for this lib. I faced issues on some android version and couldn't fix it.
The main purpose of this lib was to use wifi network no matter what (I'm using it to configure IOT device).
I suggest use it by default, or add variable to `connect(... Bool forceUseNetwork, ...)`  function.
I think it is good to have identical API for both, iOS and Android. In this case iOS will ignore variable.

edpaget commented:
That makes sense to me. I'll update it to do that. I think we'd probably still to have a function for unbinding in Android (and a corresponding dummy function in iOS), unless you think there's a better place that can be handled?

edpaget commented:
Change to make bind and unbind boolean arguments that get passed to the remove and connect methods, and also updated the ios files to accept dummy arguments. 

I also created wrapper javascript functions, to make sure the interface didn't change for current users (two arguments to `connect` will be interpreted as an `ssid` and `callback`), and added TypeScript type defs since I'm using that, but can definitely back either of those changes out if you'd prefer.

tadasr commented:
@edpaget that looks just great. There are two thing I think would be great to add:
1. If we are adding mid-layer in JS, let's add all functions there (`isAvaliable` and `getSSID`).
2. Can we update example with latest functionality?

saurabytes commented on android/src/main/java/com/tadasr/IOTWifi/IOTWifiModule.java:
```diff
@@ -97,9 +104,66 @@ private void connectToWifi(String ssid, String passphrase, Boolean isWEP, Callba
     }
 
+    private void bindToNetwork(final String ssid) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
```
I think fast fail should be preferred over bing enclosing If conditions. Reverse the condition
```
if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
 return;
}
```

saurabytes commented on android/src/main/java/com/tadasr/IOTWifi/IOTWifiModule.java:
```diff
@@ -88,6 +92,9 @@ private void connectToWifi(String ssid, String passphrase, Boolean isWEP, Callba
             }
             try {
                 Thread.sleep(3000);
+                if (bindNetwork) {
```
Any specific reason you are binding after the sleep?

edpaget replied:
I assumed the sleep was added to wait for the network to be connected and available, originally this was not checking that the bound network has the correct SSID, but now that it is that bind could happen elsewhere.

saurabytes commented on android/src/main/java/com/tadasr/IOTWifi/IOTWifiModule.java:
```diff
@@ -97,9 +104,66 @@ private void connectToWifi(String ssid, String passphrase, Boolean isWEP, Callba
+                                    connectivityManager.bindProcessToNetwork(null);
+                                } else {
+                                    ConnectivityManager.setProcessDefaultNetwork(null);
+                                }
```
Mind extracting that into one common method taking a `network` param
Above If..else block seems to be repeated from lines 124 - 128 and then 153 - 157.

saurabytes commented on android/src/main/java/com/tadasr/IOTWifi/IOTWifiModule.java:
```diff
@@ -97,9 +104,66 @@ private void connectToWifi(String ssid, String passphrase, Boolean isWEP, Callba
+            connectivityManager.bindProcessToNetwork(null);
+        } else {
+            ConnectivityManager.setProcessDefaultNetwork(null);
+        }
```
Re-Refactor comment above

saurabytes commented on android/src/main/java/com/tadasr/IOTWifi/IOTWifiModule.java:
```diff
@@ -123,6 +187,7 @@ public void removeSSID(String ssid) {
     public void getSSID(Callback callback) {
         WifiInfo info = wifiManager.getConnectionInfo();
         String ssid = info.getSSID();
+        Log.d("wifi", ssid);
```
Add more to this log. Something like `Current wifi SSID: `

edpaget replied:
Am just going to remove the logging statement.

saurabytes commented on index.d.ts:
```diff
@@ -0,0 +1,20 @@
+  export namespace RNWifi {
+    function isAvailable(cb: (available: boolean) => void): void;
+    function getSSID(cb: (ssid: string) => void): void;
+    function connect(...args: RemoveConnectArgs): void;
```
Same `RemoveConnectArgs` used for connect method. Creating a separate type would be good for maintenance

saurabytes commented:
I tried to fix as many issues as possible with above PR but there is still one left. For some reason neither example app nor my other app are able to make `isAvailable` call. I dont need it for my app but it would be nice to see if somebody get figure out the issue and fix it before merge.
Note: I did update the code to fix the type `isAvaliable` to `isAvailable`

saurabytes commented:
Update: I have fixed following issues
```

- Change isAvailable to isApiAvailable
- Ignores binding crashes on SDK 23 (still haven't figure out why its crashing)
- Fixes the iOS react exports to match the index.js and fixes the crash
- Fix: detecting pre-added system config which can't be enabled or removed

```
 in this PR https://github.com/edpaget/react-native-iot-wifi/pull/2, can you review?

saurabytes commented:
With the fixes merged above I think the PR is is much better shape and supports following
- Optional Force binding to a Wifi on both iOS and Android platforms
- Android: Better error handling
- Android: Detection for a system added config which cant be reused, removed or updated
It does do a lot of other cleanup and re-use of code

I have tested this on multiple API levels and iOS version 12

saurabytes commented on android/src/main/java/com/tadasr/IOTWifi/IOTWifiModule.java:
```diff
@@ -13,47 +16,104 @@
-    
+
+    private String errorFromCode(int errorCode) {
+        return "Error: " + errorCode;
```
I should have used `ErrorCode`, can you update it please?

edpaget replied:
I assume you mean it should just be capitalized?

saurabytes replied:
I meant
`"Error: " + errorCode`; 
should be 
`"ErrorCode: " + errorCode;`

saurabytes replied:
errorCode is an argument to the java method so it shouldn't be capitalized. Sorry about the confusion.

edpaget replied:
Got it I will change that around. i thought the request was odd.

edpaget commented:
Sorry let this task slip. I've updated the Error -> ErrorCode

saurabytes commented:
@tadasr I think this PR is in good shape now. Feel free to use my https://github.com/tadasr/react-native-iot-wifi/pull/10#issuecomment-485881032 for release notes. I might also suggest a major version upgrade for the isApiAvialable api change.

tadasr commented:
Thanks @saurabytes & @edpaget!
